feat: Slow updates experimentation

LHerskind opened the pull request:
Fixes #2801. (Also fixes #3214)

Experimentation with slow updates in Noir. 

- Adds a slow update structure
  - Sparse tree to support membership and non-memberships
- Wraps this structure in a contract to make interaction easier
- Adds an oracle to make it practical to pass the >200 params.
- Implement token contract with blacklist and access control in the slow tree.



From https://github.com/AztecProtocol/aztec-packages/pull/2878:
For the experimentation here needed a name for my "data-stack" as it could practically hold whatever you want to throw at it. Was listening to https://www.youtube.com/watch?v=DTrNwbemNK0 while building so its the pez dispenser. The PEZ dispenser is essentially a stack managed by the PXE that hold mints (PEZ Dispenser = "PfeffErminZ"/peppermint dispenser). A mint is Fr[] which can hold whatever data really, here just used to feed membership proofs and update proofs into the slow updates map.

# Checklist:
Remove the checklist to signal you've completed it. Enable auto-merge if the PR is ready to merge.
- [ ] If the pull request requires a cryptography review (e.g. cryptographic algorithm implementations) I have added the 'crypto' tag.
- [ ] I have reviewed my diff in github, line by line and removed unexpected formatting changes, testing logs, or commented-out code.
- [ ] Every change is related to the PR description.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this pull request to relevant issues (if any exist).

LHerskind committed:
feat: initial impl

LHerskind committed:
chore: minor fixes

LHerskind committed:
feat: initial dirty slow_inclusion map

LHerskind committed:
chore: fixed size map

LHerskind committed:
chore: fix tsconfig.json

LHerskind committed:
chore: purge slow_struct

LHerskind committed:
feat: generic size. Simple full size sparse tree

LHerskind committed:
chore: cleanup

LHerskind committed:
chore: update cli_docs_sandbox_test

LHerskind committed:
chore: update comment

LHerskind committed:
chore: minor cleanup

benesjan commented on yarn-project/aztec-nr/slow-updates-tree/src/slow_map.nr:
```diff
@@ -0,0 +1,253 @@
+global EPOCH_LENGTH: u120 = 100;
+
+fn compute_next_change(time: Field) -> Field {
+  (((time as u120 / EPOCH_LENGTH + 1) * EPOCH_LENGTH)) as Field
```
Why is there the +1 in the denominator? Is it to prevent some rounding issue?

LHerskind replied:
It is not in the denominator, it is just looking like garbage hehe, it is practically
```
((time / EPOCH_LENGTH) + 1) * EPOCH_LENGTH
```
So it is just the time of the "next" epoch.

benesjan commented on yarn-project/noir-contracts/src/contracts/slow_tree_contract/src/main.nr:
```diff
@@ -0,0 +1,180 @@
+    };
+
+    global TREE_HEIGHT: Field = 254;
+    global MEMBERSHIP_SIZE: Field = 256; // TREE_HEIGTH + 2
```
```suggestion
    global MEMBERSHIP_SIZE: Field = 256; // TREE_HEIGHT + 2
```

Ser, did you not enable the spell check extension? You are facing a typo Karen here
<img width="381" alt="image" src="https://github.com/AztecProtocol/aztec-packages/assets/13470840/f461ad35-e390-49fe-b901-25ce8f4b5d31">

benesjan commented on yarn-project/aztec-nr/slow-updates-tree/src/slow_map.nr:
```diff
@@ -0,0 +1,253 @@
+  }
+
+  // A variation of `update_at` that skips the merkle-membership checks.
+  // To be used by a contract wich has already checked the merkle-membership.
```
```suggestion
  // To be used by a contract which has already checked the merkle-membership.
```

benesjan reviewed (commented):
Looks good, but would not merge until the slow updates tree e2e is reliably passing and enabled in CI.

rahul-kothari commented on yarn-project/end-to-end/src/e2e_slow_tree.test.ts:
```diff
@@ -0,0 +1,146 @@
+
+    const zeroProof = await getMembershipProof(key, false);
+    logger(`"Reads" tree[${zeroProof.index}] from the tree, equal to ${zeroProof.value}`);
+    await contract.methods.read_at(zeroProof).send().wait();
```
why do you not need to pass in `value` here but in L103 you do?

LHerskind replied:
In L103 I was just overwriting the value to make it explicit.

rahul-kothari commented on yarn-project/end-to-end/src/e2e_slow_tree.test.ts:
```diff
@@ -0,0 +1,146 @@
+    logger(`"Reads" tree[${zeroProof.index}] from the tree, equal to ${zeroProof.value}`);
+    await contract.methods.read_at(zeroProof).send().wait();
+
+    // Progress time to beyond the update and thereby commit it to the tree.
```
```suggestion
    // Progress time to beyond the epoch and thereby commit it to the tree.
```

rahul-kothari commented on yarn-project/end-to-end/src/e2e_slow_tree.test.ts:
```diff
@@ -0,0 +1,146 @@
+
+    const db = levelup(createMemDown());
+    const depth = 254;
+    const tree = await newTree(SparseTree, db, hasher, 'test', depth);
```
```suggestion
    // a simulator to help keep track of the state of the expected tree and generate sibling paths
    const tree = await newTree(SparseTree, db, hasher, 'test', depth);
```

rahul-kothari commented on yarn-project/end-to-end/src/e2e_slow_tree.test.ts:
```diff
@@ -0,0 +1,146 @@
+    logger(
+      `Tries to "read" tree[${zeroProof.index}] from the tree, but is rejected as value is not ${zeroProof.value}`,
+    );
+    await expect(contract.methods.read_at(zeroProof).simulate()).rejects.toThrowError(
```
I remember your GH issue spoke about needing non-membership checks. This is it right? Or is there more?

LHerskind replied:
With the sparse tree you can just store 0 or 1 at the leaf to have inclusions and check the value 👍

rahul-kothari commented on yarn-project/end-to-end/src/e2e_slow_tree.test.ts:
```diff
@@ -0,0 +1,146 @@
+      .wait();
+    await tree.updateLeaf(new Fr(4).toBuffer(), key);
+
+    await status(key);
```
should this require any expects?

rahul-kothari commented on yarn-project/noir-contracts/src/contracts/slow_tree_contract/src/main.nr:
```diff
@@ -0,0 +1,180 @@
+    }
+
+    #[aztec(public)]
+    fn update_at(p: SlowUpdateProof<TREE_HEIGHT, UPDATE_SIZE>) {
```
Use `update_at_pub()` since everywhere in the library you explicitly add "_pub"

LHerskind commented on yarn-project/end-to-end/src/e2e_slow_tree.test.ts:
```diff
@@ -0,0 +1,146 @@
+  /**
+   * Pedersen hasher for the slow tree to match noir hashing.
+   */
+  class PedersenHasher implements Hasher {
```
Different pedersen back then I think 💀

LHerskind committed:
chore: update pedersen 💀

LHerskind committed:
feat: add oracle for slow updates (#2878)

Experimental, very much in progress.

Fixes #2852, building on top of #2732. 

For the experimentation here needed a name for my "data-stack" as it
could practically hold whatever you want to throw at it. Was listening
to https://www.youtube.com/watch?v=DTrNwbemNK0 while building so its the
pez dispenser. The PEZ dispenser is essentially a stack managed by the
PXE that hold `mint`s (PEZ Dispenser = "PfeffErminZ"/peppermint
dispenser). A `mint` is `Fr[]` which can hold whatever data really, here
just used to feed membership proofs and update proofs into the slow
updates map.

LHerskind committed:
chore: remove allo

LHerskind committed:
feat: initial usdc-like token

LHerskind committed:
chore: artifact inclusion

LHerskind committed:
test: add blacklist tests

LHerskind committed:
chore: add to ci

LHerskind committed:
feat: use slow updates for almost all access control in token

LHerskind committed:
chore: 🧹

LHerskind committed:
chore: fix import

LHerskind committed:
chore: add extra accounts if needed

LHerskind committed:
chore: add slow-tree test to ci

LHerskind committed:
chore: recompile noir

LHerskind committed:
chore: noir updates

LHerskind committed:
chore: fix imports

benesjan commented on yarn-project/noir-contracts/src/contracts/token_blacklist_contract/src/main.nr:
```diff
@@ -0,0 +1,501 @@
+        amount: Field,
+        nonce: Field,
+    ) -> Field {
+        let slow = SlowMap::at(AztecAddress::new(storage.slow_update.get_note().value));
```
I think it would be clearer if the interface was called SlowTree given that it is calling the slow tree contract. It confused me for a bit when I instinctively navigated to read_at functionality in SlowMap struct and I was confused why public context was used there (given that bellow it's calling a private func).

benesjan commented on yarn-project/pxe/src/database/memory_db.ts:
```diff
@@ -44,11 +45,20 @@ export class MemoryDB extends MemoryContractDatabase implements Database {
     return Promise.resolve();
   }
 
+  public addMint(mint: Fr[]): Promise<void> {
```
Maybe pushSlow?

rahul-kothari replied:
I'd go for `pushToSlowUpdatesTree()` to make it easy for external devs

benesjan commented on yarn-project/pxe/src/database/memory_db.ts:
```diff
@@ -44,11 +45,20 @@ export class MemoryDB extends MemoryContractDatabase implements Database {
+    return Promise.resolve();
+  }
+
+  public popMint(): Promise<Fr[] | undefined> {
```
And popSlow?

benesjan reviewed (changes requested):
Great job getting this to work. My brain almost melted when reviewing this.

Will approve once the naming and maybe the slow tree test gets improved.

rahul-kothari commented on yarn-project/types/src/interfaces/pxe.ts:
```diff
@@ -42,6 +42,12 @@ export interface PXE {
   addAuthWitness(authWitness: AuthWitness): Promise<void>;
 
+  /**
+   * Adding a mint to the pez dispenser.
```
Consider using the full form of PEZ - external folks look at it as docs
Also as @benesjan highlighted, add/popMint() could be changed to addTo/popFromSlowUpdatesTree()?

Not sure if `mint` is the right term here as well

rahul-kothari commented on yarn-project/pxe/src/database/memory_db.ts:
```diff
@@ -20,6 +20,7 @@ export class MemoryDB extends MemoryContractDatabase implements Database {
   private globalVariablesHash: Fr | undefined;
   private addresses: CompleteAddress[] = [];
   private authWitnesses: Record<string, Fr[]> = {};
+  private pezDispenser: Fr[][] = [];
```
+1

rahul-kothari commented on yarn-project/pxe/src/database/database.ts:
```diff
@@ -24,6 +24,18 @@ export interface Database extends ContractDatabase {
+   * Get the next mint from the pez dispenser.
+   * @returns A promise that resolves to an array of field elements representing the mint.
+   */
+  popMint(): Promise<Fr[] | undefined>;
```
do we want to throw or just return undefined?

rahul-kothari commented on yarn-project/aztec-nr/slow-updates-tree/Nargo.toml:
```diff
@@ -0,0 +1,8 @@
+[package]
+name = "slow_updates_tree"
+authors = ["aztec-labs"]
+compiler_version = "0.1"
```
compiler version is enforced now. Maybe do `>=0.18.0`

rahul-kothari commented on yarn-project/aztec-nr/slow-updates-tree/src/slow_map.nr:
```diff
@@ -0,0 +1,271 @@
+  // Should only be used when updates from public are desired, since the hashing will be
+  // costly since done by sequencer.
+  pub fn update_at(self: Self, p: SlowUpdateProof<N, M>) {
+    // TODO: Constrain the index to be within the tree.
```
does this need a new issue?

rahul-kothari commented on yarn-project/aztec-nr/slow-updates-tree/src/slow_map.nr:
```diff
@@ -0,0 +1,271 @@
+  }
+}
+
+pub fn compute_merkle_root<N>(leaf: Field, index: Field, hash_path: [Field; N]) -> Field {
```
damn I thought noir had a library for this

rahul-kothari commented on yarn-project/end-to-end/src/e2e_slow_tree.test.ts:
```diff
@@ -0,0 +1,131 @@
+      };
+    };
+
+    const getMembershipMint = (proof: { index: bigint; value: Fr; sibling_path: Fr[] }) => {
```
why "mint"?

rahul-kothari commented on yarn-project/end-to-end/src/e2e_slow_tree.test.ts:
```diff
@@ -0,0 +1,131 @@
+    const zeroProof = await getMembershipProof(key, false);
+    logger(`"Reads" tree[${zeroProof.index}] from the tree, equal to ${zeroProof.value}`);
+    await wallet.addMint(getMembershipMint({ ...zeroProof, value: new Fr(0) }));
+    await contract.methods.read_at(key).send().wait();
```
perhaps fetch the value and expect?

rahul-kothari commented on yarn-project/noir-contracts/src/contracts/slow_tree_contract/src/main.nr:
```diff
@@ -0,0 +1,169 @@
+
+// This contract allow us to "read" public storage in private through a delayed tree.
+// More documentation need to be outlined for this properly, but there is some in 
+// https://github.com/AztecProtocol/aztec-packages/pull/2732/files
```
delete this hahah

rahul-kothari commented on yarn-project/noir-contracts/src/contracts/slow_tree_contract/src/main.nr:
```diff
@@ -0,0 +1,169 @@
+    fn read_at(index: Field) -> Field {
+        let fields = pop_mint();
+        let p: MembershipProof<TREE_HEIGHT, MEMBERSHIP_SIZE> = deserialize_membership_proof(fields);
+        debug_log_format("Indexes: {0} {1}", [index, p.index]);
```
dangling debug stmt

rahul-kothari commented on yarn-project/noir-contracts/src/contracts/slow_tree_contract/src/main.nr:
```diff
@@ -0,0 +1,169 @@
+    }
+
+    unconstrained fn compute_note_hash_and_nullifier(_contract_address: Field, _nonce: Field, _storage_slot: Field, _preimage: [Field; 4]) -> [Field; 4] {
+        [0x0d, 0x0e, 0x0a, 0x0d]
```
LOOOOOL

rahul-kothari commented on yarn-project/noir-contracts/src/contracts/token_blacklist_contract/Nargo.toml:
```diff
@@ -0,0 +1,11 @@
+[package]
+name = "token_blacklist_contract"
+authors = [""]
+compiler_version = "0.1"
```
I believe noir now enforces compiler version

rahul-kothari commented on yarn-project/noir-contracts/src/contracts/token_blacklist_contract/src/types/balance_set.nr:
```diff
@@ -0,0 +1,126 @@
+use crate::types::token_note::{TokenNote, TOKEN_NOTE_LEN, TokenNoteMethods};
+
+// A set implementing standard manipulation of balances.
+// Does not require spending key, but only knowledge.
```
but only knowledge of what?

rahul-kothari commented on yarn-project/noir-contracts/src/contracts/token_blacklist_contract/src/types/balance_set.nr:
```diff
@@ -0,0 +1,126 @@
+                // This will call the the `compute_nullifer` function of the `token_note`
+                // which require knowledge of the secret key (currently the users encryption key).
+                // The contract logic must ensure that the spending key is used as well.
+                // docs:start:remove
```
we can remove docs references from this file since it is a copy passte

rahul-kothari commented on yarn-project/noir-contracts/src/contracts/token_blacklist_contract/src/main.nr:
```diff
@@ -0,0 +1,501 @@
+// docs:start:token_all
```
probably can remove these doc references

rahul-kothari commented on yarn-project/noir-contracts/src/contracts/token_blacklist_contract/src/main.nr:
```diff
@@ -0,0 +1,501 @@
+mod types;
+mod interfaces;
+
+// Minimal token implementation that supports `AuthWit` accounts.
```
replace this with the slow updates tree stuff

rahul-kothari commented on yarn-project/noir-contracts/src/contracts/token_blacklist_contract/src/main.nr:
```diff
@@ -0,0 +1,501 @@
+
+
+// TODO's: @LHerskind
+// We are currently unable to do constructor -> private calls
```
reference the issue you created

LHerskind committed:
chore: addressing minor comments

LHerskind committed:
chore: add constraints to e2e_slow_tree.test.ts

LHerskind committed:
refactor: use capsules and capsule dispenser instead.

LHerskind committed:
chore: formatting

LHerskind committed:
chore: merge fixing

LHerskind committed:
fix: merge

LHerskind committed:
chore: linter

benesjan commented on yarn-project/pxe/src/database/database.ts:
```diff
@@ -24,6 +24,18 @@ export interface Database extends ContractDatabase {
+  addCapsule(capsule: Fr[]): Promise<void>;
+
+  /**
+   * Get the next capsule from the capsule dispenser.
```
```suggestion
   * Get the next capsule from the capsule dispenser.
   * @remarks A capsule is a "blob" of data that is passed to the contract through an oracle.
```

benesjan commented on yarn-project/pxe/src/database/database.ts:
```diff
@@ -24,6 +24,18 @@ export interface Database extends ContractDatabase {
   getAuthWitness(messageHash: Fr): Promise<Fr[]>;
 
+  /**
+   * Adding a capsule to the capsule dispenser.
```
```suggestion
   * Adding a capsule to the capsule dispenser.
   * @remarks A capsule is a "blob" of data that is passed to the contract through an oracle.
```

Everybody would be searching what a capsule is.

benesjan commented on yarn-project/types/src/interfaces/pxe.ts:
```diff
@@ -42,6 +42,12 @@ export interface PXE {
 
+  /**
+   * Adding a capsule to the capsule dispenser.
+   * @param capsule - An array of field elements representing the capsule.
```
```suggestion
   * @param capsule - An array of field elements representing the capsule.
   * @remarks A capsule is a "blob" of data that is passed to the contract through an oracle.
```

benesjan replied:
```suggestion
   * @param capsule - An array of field elements representing the capsule.
   * @remarks A capsule is a "blob" of data that is passed to the contract through an oracle.
```

benesjan commented on yarn-project/acir-simulator/src/client/db_oracle.ts:
```diff
@@ -35,6 +35,12 @@ export interface DBOracle extends CommitmentsDB {
 
+  /**
+   * Retrieve a capsule from the capsule dispenser.
+   * @returns A promise that resolves to an array of field elements representing the capsule.
```
```suggestion
   * @returns A promise that resolves to an array of field elements representing the capsule.
   * @remarks A capsule is a "blob" of data that is passed to the contract through an oracle.
```

benesjan commented on yarn-project/acir-simulator/src/client/view_data_oracle.ts:
```diff
@@ -57,7 +57,16 @@ export class ViewDataOracle extends TypedOracle {
   /**
-   * Gets some notes for a storage slot.
+   * Pops a capsule from the capsule dispenser
+   * @returns The capsule values
```
```suggestion
   * @returns The capsule values
   * @remarks A capsule is a "blob" of data that is passed to the contract through an oracle.
```

benesjan commented on yarn-project/noir-contracts/src/contracts/slow_tree_contract/src/capsule.nr:
```diff
@@ -0,0 +1,6 @@
+#[oracle(popCapsule)]
+fn pop_capsule_oracle<N>() -> [Field; N] {}
+
+unconstrained pub fn pop_capsule<N>() -> [Field; N] {
```
```suggestion
// A capsule is a "blob" of data that is passed to the contract through an oracle.
unconstrained pub fn pop_capsule<N>() -> [Field; N] {
```

benesjan commented on yarn-project/acir-simulator/src/client/view_data_oracle.ts:
```diff
@@ -57,7 +57,16 @@ export class ViewDataOracle extends TypedOracle {
+
+  /**
+   * Gets some notes for a contract address and storage slot.
+   * Returns a flattened array containing real-note-count and note preimages.
```
```suggestion
   * Returns a flattened array containing filtered notes.
```

Why the "real-note-count"? It just returns an array of notes.

benesjan reviewed (approved):
LGTM! Just a few tsdocs improvements

LHerskind commented on yarn-project/acir-simulator/src/client/view_data_oracle.ts:
```diff
@@ -57,7 +57,16 @@ export class ViewDataOracle extends TypedOracle {
+
+  /**
+   * Gets some notes for a contract address and storage slot.
+   * Returns a flattened array containing real-note-count and note preimages.
```
Think that have come from a merge at some point 🤔 I don't recall every adding real-note-count

LHerskind committed:
chore: formatting

LHerskind committed:
chore: address comments

LHerskind committed:
chore: compiler version nargo toml
